pennylane_lightning/core/src/simulators/lightning_gpu/StateVectorCudaManaged.hpp

```diff
@@ -0,0 +1,1969 @@
+     * @param tgts Target qubits.
+     * @return auto Expectation value.
+     */
+    auto getExpectationValueDeviceMatrix(const CFP_t *matrix,
+                                         const std::vector<std::size_t> &tgts) {
+        void *extraWorkspace = nullptr;
+        size_t extraWorkspaceSizeInBytes = 0;
+
+        std::vector<int> tgtsInt(tgts.size());
+        std::transform(
+            tgts.begin(), tgts.end(), tgtsInt.begin(), [&](std::size_t x) {
+                return static_cast<int>(BaseType::getNumQubits() - 1 - x);
+            });
+
+        size_t nIndexBits = BaseType::getNumQubits();
+        cudaDataType_t data_type;
+        cudaDataType_t expectationDataType =
+            CUDA_C_64F; // Requested by the custatevecComputeExpectation API
+        custatevecComputeType_t compute_type;
+
+        if constexpr (std::is_same_v<CFP_t, cuDoubleComplex> ||
+                      std::is_same_v<CFP_t, double2>) {
+            data_type = CUDA_C_64F;
+            compute_type = CUSTATEVEC_COMPUTE_64F;
+        } else {
+            data_type = CUDA_C_32F;
+            compute_type = CUSTATEVEC_COMPUTE_32F;
+        }
+
+        // check the size of external workspace
+        PL_CUSTATEVEC_IS_SUCCESS(custatevecComputeExpectationGetWorkspaceSize(
+            /* custatevecHandle_t */ handle_.get(),
+            /* cudaDataType_t */ data_type,
+            /* const uint32_t */ nIndexBits,
+            /* const void* */ matrix,
+            /* cudaDataType_t */ data_type,
+            /* custatevecMatrixLayout_t */ CUSTATEVEC_MATRIX_LAYOUT_ROW,
+            /* const uint32_t */ tgtsInt.size(),
+            /* custatevecComputeType_t */ compute_type,
+            /* size_t* */ &extraWorkspaceSizeInBytes));
+
+        if (extraWorkspaceSizeInBytes > 0) {
+            PL_CUDA_IS_SUCCESS(
+                cudaMalloc(&extraWorkspace, extraWorkspaceSizeInBytes));
+        }
+
+        CFP_t expect;
+
+        // compute expectation
+        PL_CUSTATEVEC_IS_SUCCESS(custatevecComputeExpectation(
+            /* custatevecHandle_t */ handle_.get(),
+            /* void* */ BaseType::getData(),
+            /* cudaDataType_t */ data_type,
+            /* const uint32_t */ nIndexBits,
+            /* void* */ &expect,
+            /* cudaDataType_t */ expectationDataType,
+            /* double* */ nullptr,
+            /* const void* */ matrix,
+            /* cudaDataType_t */ data_type,
+            /* custatevecMatrixLayout_t */ CUSTATEVEC_MATRIX_LAYOUT_ROW,
+            /* const int32_t* */ tgtsInt.data(),
+            /* const uint32_t */ tgtsInt.size(),
+            /* custatevecComputeType_t */ compute_type,
+            /* void* */ extraWorkspace,
+            /* size_t */ extraWorkspaceSizeInBytes));
+
+        if (extraWorkspaceSizeInBytes)
+            PL_CUDA_IS_SUCCESS(cudaFree(extraWorkspace));
+        return static_cast<PrecisionT>(expect.x);
+    }
+};
+}; // namespace Pennylane::LightningGPU
```

Comment:
We should also plan to break up this file as it is very long. Maybe we can adapt some of the LK or LQ designs to match, or move some of the CUDA-specific API calls into separate free functions that we can call per required op.

Reply:
Moved measurements related methods to the measurement class